src/resolution/graph.rs

```diff
@@ -623,17 +623,43 @@ impl Graph {
         dependencies.insert(specifier.clone(), (*child_pkg_id).clone());
       }
 
+      // store all the required packages which can then be used to check
+      // for any optional dependencies
+      if version_info.optional_dependencies.is_empty() {
+        required_packages.extend(dependencies.values().map(|id| id.nv.clone()));
+      } else {
+        required_packages.extend(
+          dependencies
+            .keys()
+            .filter(|key| {
+              !version_info.optional_dependencies.contains_key(*key)
+            })
+            .map(|key| dependencies.get(key).unwrap().nv.clone()),
+        );
+        had_optional = true;
+      }
+
       packages.insert(
         (*pkg_id).clone(),
         NpmResolutionPackage {
           copy_index: copy_index_resolver.resolve(pkg_id),
+          optional: false,
           pkg_id: (*pkg_id).clone(),
-          dist,
+          cpu: version_info.cpu.clone(),
+          os: version_info.os.clone(),
+          dist: version_info.dist.clone(),
           dependencies,
         },
       );
     }
 
+    // only bother doing this if an optional dependency was found
+    if had_optional {
+      for package in packages.values_mut() {
+        package.optional = !required_packages.contains(&package.pkg_id.nv);
```

Comment:
I was just writing tests for this and realized this doesn't work for required dependencies of optional dependencies.